pkg/controllers/machine_controller_volumes.go

```diff
@@ -83,75 +84,153 @@ func getLastVolumeSize(machine *api.Machine, volumeID string) *int64 {
+
 		log.V(1).Info("Successfully reconciled volume", "volumeName", volume.Name, "volumeID", volumeID)
-		volumeStates = append(volumeStates, api.VolumeStatus{
-			Name:   volume.Name,
-			Handle: volumeID,
-			State:  api.VolumeStateAttached,
-			Size:   volumeSize,
-		})
 	}
 
+	newVolumeStatus := convertVolumesMapToListAndNormalize(currentStatus)
+
 	if len(errs) > 0 {
-		return nil, fmt.Errorf("attach/detach error(s): %v", errs)
+		return newVolumeStatus, fmt.Errorf("volume reconciliation error(s): %v", errs)
+	}
+	return newVolumeStatus, nil
+}
+
+func (r *MachineReconciler) deleteDetachedVolumes(ctx context.Context, log logr.Logger, status map[string]*api.VolumeStatus, mounter VolumeMounter) []error {
+	var errs []error
+
+	// Deleting detached volumes from previous reconciliation loop
+	for volumeName, volumeStatus := range status {
+		if volumeStatus.DeletedAt == nil {
+			continue
+		}
+
+		log.V(1).Info("Unmounting volume if mounted", "volumeName", volumeName)
+		if err := mounter.DeleteVolume(ctx, volumeName); err != nil && !errors.Is(err, ErrMountedVolumeNotFound) {
+			errs = append(errs, fmt.Errorf("error unmounting volume %s: %w", volumeName, err))
+			continue
+		}
+
+		delete(status, volumeName)
 	}
-	return volumeStates, nil
+
+	return errs
 }
 
-func (r *MachineReconciler) deleteVolume(ctx context.Context, log logr.Logger, mounter VolumeMounter, attacher VolumeAttacher, volumeName string) error {
-	log.V(1).Info("Detaching volume if attached")
-	if err := attacher.DetachVolume(volumeName); err != nil && !errors.Is(err, ErrAttachedVolumeNotFound) {
-		return fmt.Errorf("error detaching volume: %w", err)
+func syncStatusWithAttachedVolumes(status map[string]*api.VolumeStatus, attached map[string]AttachVolume) {
+	if len(attached) == 0 {
+		return
 	}
 
-	log.V(1).Info("Unmounting volume if mounted")
-	if err := mounter.DeleteVolume(ctx, volumeName); err != nil && !errors.Is(err, ErrMountedVolumeNotFound) {
-		return fmt.Errorf("error unmounting volume: %w", err)
+	for key, volumeStatus := range status {
+		_, ok := attached[key]
+		if !ok && volumeStatus.DeletedAt == nil {
+			volumeStatus.DeletedAt = ptr.To[time.Time](time.Now())
+		}
 	}
 
-	return nil
+	for key := range attached {
+		_, ok := status[key]
+		if ok {
+			continue
+		}
+
+		volume := attached[key]
+		volumeStatus := api.VolumeStatus{
+			Name:  volume.Name,
+			State: api.VolumeStateAttached,
+			Size:  volume.Spec.Size,
+		}
+		status[key] = &volumeStatus
+	}
+}
+
+func convertVolumesMapToListAndNormalize(currentStatus map[string]*api.VolumeStatus) []api.VolumeStatus {
+	newVolumeStatus := make([]api.VolumeStatus, 0, len(currentStatus))
+	for _, status := range currentStatus {
+		newVolumeStatus = append(newVolumeStatus, *status)
+	}
+
+	// sort is required for avoid reconcile loop triggered by different order of volumes.
+	sort.SliceStable(newVolumeStatus, func(i, j int) bool {
+		return strings.Compare(newVolumeStatus[i].Name, newVolumeStatus[j].Name) == -1
```

Comment:
Could you please explain why we are not simply using `newVolumeStatus[i].Name < newVolumeStatus[j].Name`. Both cases perform lexicographic comparison if I am correct.

Reply:
fixed